client/json.go

```diff
@@ -0,0 +1,433 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package client
+
+import (
+	"encoding/json"
+	"io"
+
+	"github.com/valyala/fastjson"
+	"golang.org/x/exp/constraints"
+)
+
+// JSON represents a JSON value that can be any valid JSON type: object, array, number, string, boolean, or null.
+// It provides type-safe access to the underlying value through various accessor methods.
+type JSON interface {
+	json.Marshaler
+	// Array returns the value as a JSON array along with a boolean indicating if the value is an array.
+	// Returns nil and false if the value is not an array.
+	Array() ([]JSON, bool)
+
+	// Object returns the value as a JSON object along with a boolean indicating if the value is an object.
+	// Returns nil and false if the value is not an object.
+	Object() (map[string]JSON, bool)
+
+	// Number returns the value as a number along with a boolean indicating if the value is a number.
+	// Returns 0 and false if the value is not a number.
+	Number() (float64, bool)
+
+	// String returns the value as a string along with a boolean indicating if the value is a string.
+	// Returns empty string and false if the value is not a string.
+	String() (string, bool)
+
+	// Bool returns the value as a boolean along with a boolean indicating if the value is a boolean.
+	// Returns false and false if the value is not a boolean.
+	Bool() (bool, bool)
+
+	// IsNull returns true if the value is null, false otherwise.
+	IsNull() bool
+
+	// Value returns the value that JSON represents.
+	// The type will be one of: map[string]JSON, []JSON, float64, string, bool, or nil.
+	Value() any
+
+	// Unwrap returns the underlying value with all nested JSON values unwrapped.
+	// For objects and arrays, this recursively unwraps all nested JSON values.
+	Unwrap() any
+
+	// Marshal writes the JSON value to the writer.
+	// Returns an error if marshaling fails.
+	Marshal(w io.Writer) error
```

Comment:
question: Why did you want to support mashalling to a writer?

Reply:
I though that if someone wants to do some fancy marshaling with a complex json object he would pass around a buffer instead of calling `MarshalJSON` and appending results all the time.